### PR TITLE
IPA Action (using Shenzhen)

### PR DESCRIPTION
**Pull request by joshdholtz**
**Because** making an IPA is awesome

And added tests!! :whale: :dolphin: :pig: :dog: :cat: :mouse: 
### With an array (uses `ipa build` arguments)

``` ruby
  ipa "-w JoshIsAwesome.xcworkspace", "-s AlwaysBeScheming", "-d ../final_destination"

  ipa_path = Actions.lane_context[ Actions::SharedValues::IPA_OUTPUT_PATH ]
  dsym_path = Actions.lane_context[ Actions::SharedValues::DSYM_OUTPUT_PATH ]
```
### With an object (uses nice keys that we convert to `ipa build` arguments)
- Anything that is nil won't get converted to arguments
  - Except for `clean` and `archive` because they don't take arguments. They **will** get converted if the keys exist in the object. - Does that make sense? Otherwise i can make it use `clean` and `archive` when true, and not use it when false :question: :question: :question:  

``` ruby
ipa({
    workspace: ENV['WORKSPACE'],
    project: nil,
    configuration: nil,
    scheme: ENV['SCHEME'],
    clean: nil,
    archive: nil,
    destination: nil,
    embed: nil,
    identity: nil,
    sdk: nil,
    ipa: 'JoshIsAwesome.ipa'p
  })

  ipa_path = Actions.lane_context[ Actions::SharedValues::IPA_OUTPUT_PATH ]
  dsym_path = Actions.lane_context[ Actions::SharedValues::DSYM_OUTPUT_PATH ]
```

### CODE REVIEW

**Comment by KrauseFx** [comment]
This looks amazing, thanks. I'll take a closer look this weekend. 

Just a note: you seem to not have the latest version of the other tools installed, which results in downgrading the versions in the `Gemfile.lock`.

**Comment by joshdholtz** [comment]
Odd! I'm pretty sure I ran bundle before I did anything but I can fix that

**Comment by joshdholtz** [comment]
@KrauseFx Updated my gems now

File: lib/fastlane/actions/ipa.rb [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        else
+          build_args = params
+        end
+
+        # If no dest directory given, default to current directory
+        absolute_dest_directory ||= Dir.pwd
+
+        if Helper.is_test?
+          Actions.lane_context[SharedValues::IPA_OUTPUT_PATH] = File.join(absolute_dest_directory, "test.ipa")
+          Actions.lane_context[SharedValues::DSYM_OUTPUT_PATH] = File.join(absolute_dest_directory, "test.app.dSTM.zip")
```
Comment (KrauseFx) [inline lib/fastlane/actions/ipa.rb]: Wups, should be `DSYM`, right? Looks like there are no tests for the set `SharedValues` yet.

**Comment by KrauseFx** [comment]
Okay, this looks really great, only an update for the README is missing. 
Could you fix the one comment and update the README to show some example calls? 

I think it's time to soon move the action information into its own file (similar to deliver: https://github.com/KrauseFx/deliver/blob/master/Deliverfile.md)

**Comment by joshdholtz** [comment]
Ah, yes sir! I can do that

**Comment by KrauseFx** [comment]
This is so cool, I couldn't wait any longer to merge, sorry :wink:

File: lib/fastlane/actions/ipa.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        params.collect do |k,v|
+          v ||= ''
+          if args = ARGS_MAP[k]
+           "#{ARGS_MAP[k]} #{v}".strip
+          end
+        end.compact
+      end
+
+      def self.find_ipa_file(dir)
+        # Finds last modified .ipa in the destination directory
```
Comment (KrauseFx) [inline lib/fastlane/actions/ipa.rb]: Maybe we should make use of the change with my recent pull request: https://github.com/nomad/shenzhen/pull/196

That means, the absolute path to the IPA is shown as the last line of `shenzhen`. What do you think?

**Comment by KrauseFx** [comment]
Check out my recent commits https://github.com/KrauseFx/fastlane/commits/master which fix/improve some of the things.

**Comment by KrauseFx** [comment]
Okay, I thought about a better integration with `deliver` itself, but couldn't come up with a perfect solution yet.

Problems:

- When building the app from `fastlane`, it's not very easy, but possible to pass the path to the generated `ipa` file
- Main problem: Once you moved out building the app from `deliver` to `fastlane`, the user will not be able to run `deliver` or `deliver --beta` any more, which should still be possible in my opinion.

It's easier to provide a better integration for Crashlytics and Hockey.

**Pull request merged**